sdk/core/core-https/src/nodeHttpsClient.ts

```diff
@@ -0,0 +1,293 @@
+  const proxySettings = request.proxySettings;
+  if (proxySettings) {
+    if (!proxyAgent) {
+      const proxyAgentOptions: httpsProxyAgent.HttpsProxyAgentOptions = {
+        host: proxySettings.host,
+        port: proxySettings.port,
+        headers: request.headers.toJSON()
+      };
+      if (proxySettings.username && proxySettings.password) {
+        proxyAgentOptions.auth = `${proxySettings.username}:${proxySettings.password}`;
+      }
+      proxyAgent = new httpsProxyAgent(proxyAgentOptions);
+    }
+    return proxyAgent;
+  } else if (request.keepAlive) {
+    if (!keepAliveAgent) {
+      keepAliveAgent = new https.Agent({
+        keepAlive: true
+      });
+    }
+
+    return keepAliveAgent;
+  } else {
+    return https.globalAgent;
+  }
+}
+
+function isReadableStream(body: any): body is NodeJS.ReadableStream {
+  return body && typeof body.pipe === "function";
+}
+
+class ReportTransform extends Transform {
+  private loadedBytes = 0;
+  private progressCallback: (progress: TransferProgressEvent) => void;
+  _transform(chunk: string | Buffer, _encoding: string, callback: Function): void {
+    this.push(chunk);
+    this.loadedBytes += chunk.length;
+    this.progressCallback({ loadedBytes: this.loadedBytes });
+    callback();
+  }
+
+  constructor(progressCallback: (progress: TransferProgressEvent) => void) {
+    super();
+    this.progressCallback = progressCallback;
+  }
+}
+
+/**
+ * A HttpsClient implementation that uses Node's "https" module to send HTTPS requests.
+ */
+export class NodeHttpsClient implements HttpsClient {
+  /**
+   * Makes a request over an underlying transport layer and returns the response.
+   * @param request The request to be made.
+   */
+  public async sendRequest(request: PipelineRequest): Promise<PipelineResponse> {
+    const abortController = new AbortController();
+    let abortListener: ((event: any) => void) | undefined;
+    if (request.abortSignal) {
+      if (request.abortSignal.aborted) {
+        throw new AbortError("The operation was aborted.");
+      }
+
+      abortListener = (event: Event) => {
+        if (event.type === "abort") {
+          abortController.abort();
+        }
+      };
+      request.abortSignal.addEventListener("abort", abortListener);
+    }
+
+    if (request.timeout > 0) {
+      setTimeout(() => {
+        abortController.abort();
+      }, request.timeout);
+    }
+
+    if (request.formData) {
+      prepareFormData(request.formData, request);
+    }
+
+    if (!request.skipDecompressResponse) {
+      request.headers.set("Accept-Encoding", "gzip,deflate");
+    }
+
+    let body = request.body;
+
+    if (body && request.onUploadProgress) {
+      const onUploadProgress = request.onUploadProgress;
+      const uploadReportStream = new ReportTransform(onUploadProgress);
+      if (isReadableStream(body)) {
+        body.pipe(uploadReportStream);
+      } else {
+        uploadReportStream.end(body);
```

Comment:
This is interesting. So `transform` only gets called once the data is read by the http client then right? Neat trick!

The only downside I see now is that we're never setting the content-length if onUploadProgress is set.

Reply:
Yeah, maybe we should do the length detection first?

Reply:
Moved it up